Riot/AppDelegate.m

```diff
@@ -1076,23 +1078,56 @@ - (void)registerUserNotificationSettings
-        quickReply.activationMode = UIUserNotificationActivationModeBackground;
-        quickReply.authenticationRequired = true;
-        quickReply.behavior = UIUserNotificationActionBehaviorTextInput;
-
-        UIMutableUserNotificationCategory* quickReplyCategory = [[UIMutableUserNotificationCategory alloc] init];
-        quickReplyCategory.identifier = @"QUICK_REPLY";
-        [quickReplyCategory setActions:@[quickReply] forContext:UIUserNotificationActionContextDefault];
-        [notificationCategories addObject:quickReplyCategory];
-
-        // Registration on iOS 8 and later
-        UIUserNotificationSettings *settings = [UIUserNotificationSettings settingsForTypes:(UIUserNotificationTypeBadge | UIUserNotificationTypeSound |UIUserNotificationTypeAlert) categories:notificationCategories];
-        [[UIApplication sharedApplication] registerUserNotificationSettings:settings];
+        if (@available(iOS 10, *)) {
+            UNTextInputNotificationAction *quickReply = [UNTextInputNotificationAction
+                    actionWithIdentifier:@"inline-reply"
+                                   title:NSLocalizedStringFromTable(@"room_message_short_placeholder", @"Vector", nil)
+                                 options:UNNotificationActionOptionAuthenticationRequired
+            ];
+
+            UNNotificationCategory *quickReplyCategory = [UNNotificationCategory
+                    categoryWithIdentifier:@"QUICK_REPLY"
+                                   actions:@[quickReply]
+                         intentIdentifiers:@[]
+                                   options:UNNotificationCategoryOptionNone];
+
+            UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+            [center setNotificationCategories:[[NSSet alloc] initWithArray:@[quickReplyCategory]]];
+            [center setDelegate:self]; // commenting this out will fall back to using the same AppDelegate methods as the iOS 9 way of doing this
+            [center requestAuthorizationWithOptions:(UNAuthorizationOptionAlert | UNAuthorizationOptionSound | UNAuthorizationOptionBadge)
+                                  completionHandler:^(BOOL granted, NSError *error)
+                                  { // code here is equivalent to self:application:didRegisterUserNotificationSettings:
+                                      if (granted) {
+                                          [self registerForRemoteNotificationsWithCompletion:nil];
+                                      }
+                                      else
+                                      {
+                                          // Clear existing token
+                                          MXKAccountManager* accountManager = [MXKAccountManager sharedManager];
+                                          [accountManager setPushDeviceToken:nil withPushOptions:nil];
+                                      }
+                                  }];
```

Comment:
We should also listen to notification settings changes that can occur while the application is running:
```
[center getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized)
    {
        [self registerForRemoteNotificationsWithCompletion:nil];
    }
    else
    {
        // Clear existing token
        MXKAccountManager* accountManager = [MXKAccountManager sharedManager];
        [accountManager setPushDeviceToken:nil withPushOptions:nil];
    }
}];
```

Reply:
When should this be triggered? Periodically? Or each time the app goes into foreground? Or on every background sync?

Reply:
Sorry you can ignore this comment I will mark it as resolved. This is something else we have to manage.
In fact notification settings could be changed at any time we have to find a way to check those changes periodically as there is no dedicated delegate method for this at this time. For example check for notification settings with in `applicationDidBecomeActive:`.